daemon/bad_make_take_values_test.go

```diff
@@ -0,0 +1,55 @@
+package daemon
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/athanorlabs/atomic-swap/coins"
+	"github.com/athanorlabs/atomic-swap/monero"
+	"github.com/athanorlabs/atomic-swap/rpcclient"
+	"github.com/athanorlabs/atomic-swap/tests"
+)
+
+// Tests end-to-end (client->swapd->client) make/take failures with currency
+// precision issues.
+func TestBadMakeTakeValues(t *testing.T) {
+	bobConf := CreateTestConf(t, tests.GetMakerTestKey(t))
+	monero.MineMinXMRBalance(t, bobConf.MoneroClient, coins.MoneroToPiconero(coins.StrToDecimal("10")))
+
+	aliceConf := CreateTestConf(t, tests.GetTakerTestKey(t))
+
+	timeout := 3 * time.Minute
+	ctx, _ := LaunchDaemons(t, timeout, bobConf, aliceConf)
+
+	bc := rpcclient.NewClient(ctx, bobConf.RPCPort)
+	ac := rpcclient.NewClient(ctx, aliceConf.RPCPort)
+
+	// Trigger a TakeOffer failure because the precision of the min/or max
+	// value combined with the exchange rate would exceed the token's precision.
+	// 14.979329 * 13.3 = 199.2250757 (7 digits of precision)
+	minMaxXMRAmt := coins.StrToDecimal("14.979329")
+	exRate := coins.StrToExchangeRate("13.3")
+	mockTether := getMockTetherAsset(t, aliceConf.EthereumClient)
+	const expectedErr = `"net_makeOffer" failed: 14.979329 XMR * 13.3 exceeds token's 6 decimal precision`
+	_, err := bc.MakeOffer(minMaxXMRAmt, minMaxXMRAmt, exRate, mockTether, false)
+	require.ErrorContains(t, err, expectedErr)
+
+	// Now configure the MakeOffer to succeed, so we can fail some TakeOffer calls
+	minXMRAmt := coins.StrToDecimal("1")
+	maxXMRAmt := coins.StrToDecimal("10")
+	providesAmt := coins.StrToDecimal("5.1234567") // 7 digits, max is 6
+	makeResp, err := bc.MakeOffer(minXMRAmt, maxXMRAmt, exRate, mockTether, false)
+	require.NoError(t, err)
+
+	// Fail because providesAmount has too much precision in the token's standard units
+	err = ac.TakeOffer(makeResp.PeerID, makeResp.OfferID, providesAmt)
+	require.ErrorContains(t, err, `"net_takeOffer" failed: "providesAmount" has too many decimal points; found=7 max=6`)
+
+	// Fail because the providesAmount has too much precision when converted into XMR
+	// 20.123456/13.3 = 1.51304[180451127819548872] (bracketed sequence repeats forever)
+	providesAmt = coins.StrToDecimal("20.123456")
+	err = ac.TakeOffer(makeResp.PeerID, makeResp.OfferID, providesAmt)
```

Comment:
I found a bug in our client RPC handling thanks to this test that you'll see later down. There is a `TakeOfferRequest` type, but no expected response type. When there was no expected response type, we weren't decoding the server's JSON-RPC response and getting returned errors. The code here was always getting `nil` even when the server returned an error. We were most likely seeing the same error when @stubbrn did the Shutdown RPC PR.